src/MICore/MIResults.cs

```diff
@@ -596,6 +596,93 @@ public override string ToString()
 
     public class MIResults
     {
+        struct Span
+        {
+            static Span _emptySpan;
+            public int Start { get; private set; }  // index first character in the substring
+            public int Length { get; private set; } // length of the substring
+            public int Extent { get { return Start + Length; } }
+            public bool IsEmpty { get { return Length == 0; } }
+            public static Span Empty { get { return _emptySpan; } }
+
+            static Span()
+            {
+                _emptySpan = new Span(0, 0);
+            }
+
+            public Span(string s)
+            {
+                Start = 0;
+                Length = s.Length;
+            }
+            public Span(int start, int len)
+            {
+                if (start < 0)
+                {
+                    throw new ArgumentException("start");
+                }
+                Start = start;
+                Length = len;
+            }
+            public Span Advance(int len)
+            {
+                if (len > Length)
+                {
+                    throw new ArgumentException("len");
+                }
+                return new Span(Start + len, Length - len);
+            }
+            public Span AdvanceTo(int pos)
+            {
+                if (Start > pos || pos > Start + Length)
+                {
+                    throw new ArgumentException("pos");
+                }
+                return new Span(pos, Length - (pos - Start));
+            }
+            public Span Prefix(int len)
+            {
+                if (len > Length)
+                {
+                    throw new ArgumentException("len");
+                }
+                return new Span(Start, len);
+            }
+            public string Extract(string theString)
+            {
+                if (Extent > theString.Length)
+                {
+                    throw new ArgumentException("theSpan");
+                }
+                return theString.Substring(Start, Length);
+            }
+            public int IndexOf(string theString, char c)
+            {
+                int i = theString.IndexOf(c, Start);
+                if (i < 0)
+                {
+                    return i;
+                }
+                return i - Start;   // Span relative offset
```

Comment:
Not sure if this will actually matter based on how you are using this class, but there should really should also be:

```csharp
if (i >= this.Extent)
{
    return -1;
}
```